[FLINK-17773] Update documentation for new WatermarkGenerator/WatermarkStrategies (#12253)
apache/flink: docs/dev/event_timestamps_watermarks.md

under the License.
-->

In this section you will learn about the APIs that Flink provides for working
with **event time** timestamps and watermarks.  For an introduction to *event
time*, *processing time*, and *ingestion time*, please refer to the
[introduction to event time]({{ site.baseurl }}/dev/event_time.html).

* toc
{:toc}

## Introduction to Watermark Strategies

> In order to work with *event time*, Flink needs to know the events'

Reviewer: ```suggestion
In order to work with *event time*, Flink needs to know the events
```

*timestamps*, meaning each element in the stream needs to have its event
timestamp *assigned*. This is usually done by accessing/extracting the
timestamp from some field in the element by using a `TimestampAssigner`.

Timestamp assignment goes hand-in-hand with generating watermarks, which tell

> the system about progress in event time. You can configure this by speciying a

Reviewer: ```suggestion
the system about progress in event time. You can configure this by specifying a
```

`WatermarkGenerator`.

> The Flink API expects both `TimestampAssigner` and `WatermarkGenerator` to be
> bundled in a `WatermarkStrategy`. As you will see below, you usually don't have
> to do this by hand because you will mostly use the convenience helper
> `WatermarkStrategies`. Here is the interface for completess' sake:

Reviewer: I think it always makes sense to let users know there is an easy option first and then outline the low level interface. Its less scary 👻 
```suggestion
The Flink API expects a `WatermarkStrategy` that contains both a `TimestampAssigner` and `WatermarkGenerator`.
A number of common strategies out of the box, available in the `WatermarkStrategies` helper, but users can also build their own strategies when required. 
Here is the interface for completeness' sake:
```

     * event timestamps, or to emit a watermark based on the event itself.
     */
    void onEvent(T event, long eventTimestamp, WatermarkOutput output);

    /**
     * Called periodically, and might emit a new watermark, or not.
     *
     * <p>The interval in which this method is called and Watermarks are generated
     * depends on {@link ExecutionConfig#getAutoWatermarkInterval()}.
     */
    void onPeriodicEmit(WatermarkOutput output);
}
{% endhighlight %}

There are two different styles of watermark generation: *periodic* and
*punctuated*.

A periodic generator usually observes to the incoming events via `onEvent()`
and then emits a watermark when the framework calls `onPeriodicEmit()`.

A puncutated generator will look at events in `onEvent()` and wait for special
*marker events* or *punctuations* that carry watermark information in the
stream. When it sees one of these events it emits a watermark immediately.
Usually, punctuated generators don't emit a watermark from `onPeriodicEmit()`.

We will look at how to implement generators for each style next.

### Writing a Periodic WatermarkGenerator

> As mentioned above, a periodic generator observes stream events and generates

Reviewer: You use the phrase "as mentioned above" alot. 
```suggestion
A periodic generator observes stream events and generates
```

    val maxTimeLag = 5000L // 5 seconds

    override def onEvent(element: MyEvent, eventTimestamp: Long): Unit = {
        // don't need to do anything because we work on processing time
    }

    override def onPeriodicEmit(): Unit = {
        output.emitWatermark(new Watermark(System.currentTimeMillis() - maxTimeLag));
    }
}
{% endhighlight %}
</div>
</div>

### Writing a Punctuated WatermarkGenerator

> As mentioned above, a punctuated watermark generator will observe the stream of

Reviewer: ```suggestion
A punctuated watermark generator will observe the stream of
```

## Generating Watermarks and Timestamps Directly in a Source

Stream sources can directly assign timestamps to the elements they produce, and
they can also emit watermarks.  When this is done, no `WatermarkStrategy` is
needed.  Note that if a timestamp assigner is used, any timestamps and
watermarks provided by the source will be overwritten.

To assign a timestamp to an element in the source directly, the source must use
the `collectWithTimestamp(...)` method on the `SourceContext`. To generate
watermarks, the source must call the `emitWatermark(Watermark)` function.

Below is a simple example of a *(non-checkpointed)* source that assigns
timestamps and generates watermarks:

<div class="codetabs" markdown="1">
<div data-lang="java" markdown="1">
{% highlight java %}
@Override
public void run(SourceContext<MyType> ctx) throws Exception {
    while (/* condition */) {
        MyType next = getNext();
        ctx.collectWithTimestamp(next, next.getEventTimestamp());

        if (next.hasWatermarkTime()) {
            ctx.emitWatermark(new Watermark(next.getWatermarkTime()));
        }
    }

> }

Reviewer: Should this use the new source interface?

Author: For this one, I just copied the existing example. I might update to the new Source interface as a follow-up, but that's very different. Maybe we shouldn't even mention this here?

Reviewer: Yeah, let’s remove it. How many people are really building custom sources anyway.

{% endhighlight %}
</div>
<div data-lang="scala" markdown="1">
{% highlight scala %}
override def run(ctx: SourceContext[MyType]): Unit = {
    while (/* condition */) {
        val next: MyType = getNext()
        ctx.collectWithTimestamp(next, next.eventTimestamp)

        if (next.hasWatermarkTime) {
            ctx.emitWatermark(new Watermark(next.getWatermarkTime))
        }
    }
}
{% endhighlight %}
</div>
</div>

> ## How Operators Are Processing Watermarks

Reviewer: ```suggestion
## How Operators Process Watermarks
```

As a general rule, operators are required to completely process a given
watermark before forwarding it downstream. For example, `WindowOperator` will

> first evaluate which windows should be fired, and only after producing all of

Reviewer: ```suggestion
first evaluate all windows that should be fired, and only after producing all of
```
